File: app/src/main/java/com/druger/refresher/fragments/DoneTaskFragment.java
package com.druger.refresher.fragments;


import android.app.Fragment;
import android.content.Context;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.druger.refresher.R;
import com.druger.refresher.adapters.DoneTaskAdapter;
import com.druger.refresher.database.DBHelper;
import com.druger.refresher.models.ModelTask;

import java.util.ArrayList;
import java.util.List;


/**
 * A simple {@link Fragment} subclass.
 */
public class DoneTaskFragment extends TaskFragment {

    public DoneTaskFragment() {
        // Required empty public constructor
    }

    OnTaskRestoreListener onTaskRestoreListener;

    public interface OnTaskRestoreListener {
        void onTaskRestore(ModelTask task);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        try {
            onTaskRestoreListener = activity;
        } catch (ClassCastException e) {
            throw new ClassCastException(activity.toString()
                    + " must implement OnTaskRestoreListener");
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_done_task, container, false);
        setupRecycler(rootView);
        return rootView;
    }

    private void setupRecycler(View rootView) {
        recyclerView = rootView.findViewById(R.id.rvDoneTasks);
        layoutManager = new LinearLayoutManager(getActivity());
        recyclerView.setLayoutManager(layoutManager);

        tasksAdapter = new DoneTaskAdapter(this);
        recyclerView.setAdapter(tasksAdapter);
    }


    @Override
    public void addTask(ModelTask newTask, boolean saveToDB) {
        int position = -1;

        for (int i = 0; i < tasksAdapter.getItemCount(); i++) {
            if (tasksAdapter.getItem(i).isTask()){
                ModelTask task = (ModelTask) tasksAdapter.getItem(i);
                if (newTask.getDate() < task.getDate()){
                    position = i;
                    break;
                }
            }
        }

        if (position != -1){
            tasksAdapter.addItem(position, newTask);
        } else {
            tasksAdapter.addItem(newTask);
        }

        if (saveToDB){
            activity.dbHelper.saveTask(newTask);
        }
    }

    @Override
    public void findTasks(String title) {
        tasksAdapter.removeAllItems();
        List<ModelTask> tasks = new ArrayList<>();
        tasks.addAll(activity.dbHelper.query().getTasks(DBHelper.SELECTION_LIKE_TITLE + " AND "
                        + DBHelper.SELECTION_STATUS,
                new String[]{"%" + title + "%", Integer.toString(ModelTask.STATUS_DONE)},
                DBHelper.TASK_DATE_COLUMN));
        for (int i = 0; i < tasks.size(); i++) {
            addTask(tasks.get(i), false);
        }
    }

    @Override
    public void checkAdapter() {
        if (tasksAdapter == null){
            tasksAdapter = new DoneTaskAdapter(this);
            addTaskFromDB();
        }
    }

    @Override
    public void addTaskFromDB() {
        tasksAdapter.removeAllItems();
        List<ModelTask> tasks = new ArrayList<>();
        tasks.addAll(activity.dbHelper.query().getTasks(DBHelper.SELECTION_STATUS,
                new String[]{Integer.toString(ModelTask.STATUS_DONE)},
                DBHelper.TASK_DATE_COLUMN));
        for (int i = 0; i < tasks.size(); i++) {
            addTask(tasks.get(i), false);
        }
    }

    @Override
    public void moveTask(ModelTask task) {
        if (task.getDate() != 0){
            alarmHelper.setAlarm(task);
        }
        if (onTaskRestoreListener != null) {
            onTaskRestoreListener.onTaskRestore(task);
        }
    }
}

File: app/src/main/java/com/druger/refresher/fragments/DoneTaskFragment.kt
package com.druger.refresher.fragments


import android.content.Context
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.druger.refresher.R
import com.druger.refresher.adapters.DoneTaskAdapter
import com.druger.refresher.models.ModelTask


/**
 * A simple {@link Fragment} subclass.
 */
class DoneTaskFragment : TaskFragment() {

    lateinit var onTaskRestoreListener: OnTaskRestoreListener

    interface OnTaskRestoreListener {
        fun onTaskRestore(task: ModelTask)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        try {
            onTaskRestoreListener = activity
        } catch (e: ClassCastException) {
            throw ClassCastException(activity.toString()
                    + " must implement OnTaskRestoreListener")
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup,
                              savedInstanceState: Bundle): View {
        val rootView = inflater.inflate(R.layout.fragment_done_task, container, false)
        setupRecycler(rootView)
        return rootView
    }

    private fun setupRecycler(rootView: View) {
        recyclerView = rootView.findViewById(R.id.rvDoneTasks)
        layoutManager = LinearLayoutManager(getActivity())
        recyclerView.layoutManager = layoutManager

        tasksAdapter = DoneTaskAdapter(this)
        recyclerView.adapter = tasksAdapter
    }

    override fun addTask(newTask: ModelTask, saveToDB: Boolean) {
        var position = -1

        for (i in 0..tasksAdapter.itemCount) {
            if (tasksAdapter.getItem(i).isTask) {
                val task = tasksAdapter.getItem(i) as ModelTask
                if (newTask.date < task.date) {
                    position = i
                    break
                }
            }
        }

        if (position != -1) {
            tasksAdapter.addItem(position, newTask)
        } else {
            tasksAdapter.addItem(newTask)
        }

        if (saveToDB) {
            activity.realmHelper.saveTask(newTask)
        }
    }

    override fun findTasks(title: String) {
        tasksAdapter.removeAllItems()
        val tasks: MutableList<ModelTask> = ArrayList()
        tasks.addAll(activity.realmHelper.getTasksByTitleAndDoneStatus(title))
        for (i in 0..tasks.size) {
            addTask(tasks.get(i), false)
        }
    }

    override fun checkAdapter() {
        if (tasksAdapter == null) {
            tasksAdapter = DoneTaskAdapter(this)
            addTaskFromDB()
        }
    }

    override fun addTaskFromDB() {
        tasksAdapter.removeAllItems()
        val tasks: MutableList<ModelTask> = ArrayList()
        tasks.addAll(activity.realmHelper.getTasksByDoneStatus())
        for (i in 0..tasks.size) {
            addTask(tasks.get(i), false)
        }
    }

    override fun moveTask(task: ModelTask) {
        if (task.date.toInt() != 0) {
            alarmHelper.setAlarm(task)
        }
        onTaskRestoreListener.onTaskRestore(task)
    }
}